gist-logs: add proper documentation, improve error handling

jamandbees opened the pull request:
The gist-logs help behaved differently when called in the following ways:

brew gist-logs
brew gist-logs --help

The former displayed a help message, the latter did _not_. I looked at the homebrew help system and saw that the gist-logs developer had implemented his own help, not using the system that other subcommands use. 

This pull request changes "brew gist-logs" to behave like "brew install". The only questionable part of the change is that I no longer set Homebrew.failed = 1, but I believe this is a side-effectless change.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
No real way to test this.
- [x] Have you successfully run `brew tests` with your changes locally?

UniqMartin commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,3 +1,5 @@
+#: usage: brew gist-logs [--new-issue|-n] <formula>
```
Can you expand on this and make it more like those special documentation comments for other commands, e.g. [`deps`](https://github.com/Homebrew/brew/blob/504152038cb591a661721b09e989f428bbc6ae0a/Library/Homebrew/cmd/deps.rb#L1-L36), i.e. explain the purpose and document the options/arguments? What you see there is basically a snippet of Markdown as understood by `ronn`. To get the correct output, both the indentation and the formatting need to be matched exactly.

You'll need to check that the output of `brew help deps` and `brew help gist-logs` looks similar in terms of formatting. As a further step, you'll need to run `brew man` to regenerate the man page and then check the formatting of `man brew`.

UniqMartin commented:
Thanks for picking this up! It needs a bit more work though (see my code comment).

> The only questionable part of the change is that I no longer set Homebrew.failed = 1, but I believe this is a side-effectless change.

That's actually not an issue, as you replaced this with raising an exception and this will also make `brew` exit with an error code. Thus your change is correct and won't alter the behavior (aside from possibly printing a much better help message).

jamandbees commented:
@UniqMartin I've made the changes you suggested; I'm not super familiar with gist-logs, but I think I've captured what it does and what its one option does.

When I generated the new man page, the man page looks fine in the man pager, and the html version looks fine in a browser, but it looks like I've got a different version of whichever gem generates the html version. You'll see in the diff that there's some (semantically insignificant) whitespace introduced, and some colons and whatnot introduced. If you'd prefer I regenerate the html man page, can you let me know what gem versions I should use to generate the html how  you'd prefer it?

Thanks for the feedback!

UniqMartin commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,3 +1,14 @@
+#:  * `gist-logs` [`--new-issue`|`-n`] <formula>
```
The cause for the different rendering is not a different Gem version but a very subtle detail in this line: It needs to end with a trailing colon (`:`). Otherwise the whole list will no longer be recognized as a definition list and instead rendered as a plain list (the man page is also rendered quite differently, if you look closely).

UniqMartin commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,3 +1,14 @@
+#:  * `gist-logs` [`--new-issue`|`-n`] <formula>
+#:     Upload logs for a failed build of <formula> to a new gist.
```
Please make that “Gist” instead of “gist”.

UniqMartin commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,3 +1,14 @@
+#:     several different ways. See [SPECIFYING FORMULAE].
+#:
+#:     If `--new-issue` is passed, automatically create a new issue in the appropriate
+#:     github repository as well as creating the gist.
```
Please make that “GitHub” instead of “github” and “Gist” instead of “gist”.

UniqMartin commented on Library/Homebrew/cmd/gist-logs.rb:
```diff
@@ -1,3 +1,14 @@
+#:     Upload logs for a failed build of <formula> to a new gist.
+#:
+#:     <formula> is usually the name of the formula to install, but it can be specified
+#:     several different ways. See [SPECIFYING FORMULAE].
```
I think this needs to be `[SPECIFYING FORMULAE][]` to be recognized and rendered correctly, but hard to tell given the many spurious changes in the current version of the man page and HTML version of it.

Also: “specified several different ways” ➡️ “specified **in** several different ways”.

UniqMartin commented:
> I've made the changes you suggested; I'm not super familiar with gist-logs, but I think I've captured what it does and what its one option does.

You did. Great work so far! See my code comments for some minor adjustments and and explanation for why the regenerated files changed so drastically.

mikemcquaid commented:
Loving this so far 👍

jamandbees commented:
@UniqMartin I've made the changes you suggested. This time, I made sure to run my diff tool against the man page in /usr/local/share/man/man1/brew.1. I can see the changes I've made, but there's an extra line difference. On line 284 of the man page, the following formatting has changed:

\fBHOMEBREW_BUILD_FROM_SOURCE\fR has become HOMEBREW_BUILD_FROM_SOURCE

I changed the wording in that file from "several different ways" to "in several different ways", as that was the origination point for my copying the wording to Specifying Formulae, so I wanted to fix the same wording issue in both places I knew it existed.

It seemed like when I removed the change and regenerated the man page, the above two control characters around HOMEBREW_BUILD_FROM_SOURCE were lost. If I check out the master branch and generate the man pages from master, I see the same change to the control characters surrounding that variable. I don't see any easily discernible cause, but I don't think that change is originating from the changes I've made.

Thanks for spending so much time with me on this! I wanted to use making a help change as a stepping stone to actually working a little more on homebrew, and you've been incredibly welcoming.

UniqMartin commented:
> I've made the changes you suggested.

I'm sorry this needs another round trip, but there still a few things that need to be addressed:

-  Capitalization (you seem to have missed some comments): “gist” ➡️ “Gist”, “github” ➡️ “GitHub”.
- `[SPECIFYING FORMULAE][]`: You added the trailing `[]` but then (accidentally) reverted this change in a later commit.
- Make sure to re-generate the man page and HTML version of it by running `brew man` as the very last step (a quick inspection shows that your changes and these generated files are out of sync).

> I can see the changes I've made, but there's an extra line difference. On line 284 of the man page, the following formatting has changed:
> 
> \fBHOMEBREW_BUILD_FROM_SOURCE\fR has become HOMEBREW_BUILD_FROM_SOURCE

These are some last-minute changes from #205 that apparently didn't make it into the final commit. (It's easy to forget to re-run `brew man` as you've already learned.) I fixed this now. If you update your Homebrew via `brew update`, check out your `gist-logs-help` branch, and run `git rebase master` (this should complete without any conflicts), then your changes will be based off the latest commit to `master` and those spurious changes will be gone. (You should do this before adding the above mentioned changes.)

If you feel comfortable using `git rebase --interactive`, consider squashing your branch to a single commit and then pushing that, when you are done. Don't worry if you aren't. I can also do that for you when integrating your changes into Homebrew. No matter if you squashed your commits or not, you should be able to see the changes that will end up in this PR once you push if you run `git diff master...gist-logs-help` locally.

> I changed the wording in that file from "several different ways" to "in several different ways", as that was the origination point for my copying the wording to Specifying Formulae, so I wanted to fix the same wording issue in both places I knew it existed.

That's fine. Thanks!

> Thanks for spending so much time with me on this! I wanted to use making a help change as a stepping stone to actually working a little more on homebrew, and you've been incredibly welcoming.

I'm glad you're feeling that way. Thanks for the kind words! ❤️

jamandbees commented:
@UniqMartin You've been so patient with me, thank you!

I've squashed my changes into one commit and pushed them to GitHub. It's funny how so much work over the course of a day is actually only a few lines when squashed down.

If I've missed another detail, please don't hesitate to let me know. I think I caught everything you mentioned this time.

UniqMartin commented:
Merged in a558629654c28fff1ccfb8e9895485e0da79e167. Thank you for this contribution to Homebrew, @jamandbees! :tada:
